File: milestone 2/python_modules/module03/ex5/ft_data_stream.py
import random
import typing

def gen_event() -> typing.Generator:
    players: list = ['alice', 'bob', 'charlie', 'dylan']
    actions: list = ['run', 'eat', 'sleep', 'grab', 'move', 'climb', 'swim', 'release', 'use']
    while True:
        yield (random.choice(players), random.choice(actions))

def consume_event(event_list: list) -> typing.Generator:
    while len(event_list) > 0:
        event: tuple = random.choice(event_list)
        event_list.remove(event)
        yield event

if __name__ == "__main__":
    print("=== Game Data Stream Processor ===\n")

    stream: typing.Generator = gen_event()
    for i in range(1000):
        name, action = next(stream)
        print(f"Event {i}: Player {name} did action {action}")

    event_list: list = [next(stream) for _ in range(10)]
    print(f"\nBuilt list of {len(event_list)} events: {event_list}")

    for event in consume_event(event_list):
        print(f"Got event from list: {event}")
        print(f"Remains in list: {event_list}")

File: milestone 2/python_modules/module03/ex5/ft_data_stream.cpp
#include <iostream>
#include <vector>
#include <string>
#include <ctime>
using namespace std;

void gen_event()
{
    int i = 0;

    srand(time(0));
    vector<string> players = {
        "alice", "bob", "charlie", "dylan"
    };
    vector<string> actions = {
        "run", "eat", "sleep", "grab", "move", "climb", "swim", "release", "use"
    };

    while (i < 100)
    {
        for (string &player : players)
        {
            for (string &action : actions)
            {
                cout << "Player " << player << " performed an action: " << action << endl;
                i++;
            }
        }
    }
}

int main(void)
{
    cout << "=== Game Data Stream Processor ===\n";
    gen_event();

    return 0;
}